_posts/2015-05-07-Hyperdrive.md

```diff
@@ -0,0 +1,203 @@
+---
+title: Hyperdrive
+excerpt: A generic API Blueprint client in Swift
+layout: post
+date: 2015-05-07 16:00:00 +0100
+author: kyle
+published: true
+comments: true
+---
+
+It's extremely common to build API clients which are tightly coupled to
+implementation detail. If you look at most API clients you will
+often find hard-coded details about how an API works such as the HTTP
+methods to use, URIs and information about content types. We design our
+clients and APIs around out-of-band specifications which overly couple
+our clients to our server.
+
+## Why is this a problem?
+
+This type of client design is rarely seen as a problem, it usually isn't
+a problem immediately and only becomes a problem when you want to
+change how the API works. Since clients are normally tightly coupled
+to implementation details of an API. This often means you have to be
+extremely careful when it comes to changing your API. You will often decide
+against making certain changes because it will often break a client and
+you'd rather stick to how it works. This problem is easily avoided if we
+don't duplicate how our implementation works many times in each client.
+
+We often design our APIs by exposing implementation details, we use IDs
+in our URIs and expect our clients to know how to construct these URIs
+by knowing about our internal database ID.
+
+## API Evolvability
+
+You can't have evolvability if clients have their controls hard-coded into
+their design at deployment. Controls have to be learned on the fly to
+allow both the server and the client change independently over time.
+
+Imagine entering an API without knowing about any implementation details,
+but instead just understanding of the semantics behind the API. We can load
```

Comment:
...understanding **the semantics of the application domain** (or instead of "application" you can stick with "API")